cs3api4lab/api/lock_manager.py

```diff
@@ -0,0 +1,113 @@
+
+from cs3api4lab.auth import check_auth_interceptor
+from cs3api4lab.auth.authenticator import Auth
+from cs3api4lab.auth.channel_connector import ChannelConnector
+from cs3api4lab.logic.storage_logic import StorageLogic
+from cs3api4lab.api.cs3_user_api import Cs3UserApi
+from cs3api4lab.config.config_manager import Cs3ConfigManager
+from cs3api4lab.exception.exceptions import LockNotFoundError
+
+class LockManager:
+
+    def __init__(self, log):
+        self.log = log
+        self.config = Cs3ConfigManager().get_config()
+        self.auth = Auth.get_authenticator(config=self.config, log=self.log)
+
+        channel = ChannelConnector().get_channel()
+        auth_interceptor = check_auth_interceptor.CheckAuthInterceptor(log, self.auth)
+        intercept_channel = grpc.intercept_channel(channel, auth_interceptor)
+
+        self.cs3_api = cs3gw_grpc.GatewayAPIStub(intercept_channel)
+        self.user_api = Cs3UserApi(log)
+        self.storage_logic = StorageLogic(log)
+        self.locks_expiration_time = self.config['locks_expiration_time']
+
+    def generate_lock_entry(self):
+        user = self._get_current_user()
+        return {self.get_my_lock_name(): json.dumps({
+            "username": user.username,
+            "idp": user.id.idp,
+            "opaque_id": user.id.opaque_id,
+            "updated": time.time(),
+            "created": time.time()
+        })}
+
+    def get_my_lock_name(self):
+        user = self._get_current_user()
+        return 'lock_' + user.username + '_' + user.id.idp + '_' + user.id.opaque_id
+
+    def get_lock_path(self, file_path):
+        file_name = '.' + file_path.split('/')[-1] + '.lock'
+        return self.config['home_dir'] + '/' + file_name
+
+    def lock_file(self, file_path, endpoint):
+        self.storage_logic.set_metadata(self.generate_lock_entry(), file_path, endpoint)
+
+    def is_lock_mine(self, file_path, endpoint):
+        lock = self._get_lock(file_path, endpoint)
+        user = self._get_current_user()
+        if lock:
+            return lock['username'] == user.username #todo add more checks, ie idp
+        return False        
+
+    def is_lock_expired(self, file_path, endpoint):
+        lock = self._get_lock(file_path, endpoint)
+        if not lock:
+            return True
+        return time.time() - lock['updated'] > datetime.timedelta(seconds=self.locks_expiration_time).total_seconds()
+
+    def is_file_locked(self, file_path, endpoint):
+        lock = self._get_lock(file_path, endpoint)
+        return True if lock else False
+
+    def handle_locks_write(self, file_path, endpoint):
+        is_locked = self.is_file_locked(file_path, endpoint)
+        is_mine = self.is_lock_mine(file_path, endpoint)
+
+        if is_locked and not is_mine and not self.is_lock_expired(file_path, endpoint):
+            file_name = file_path.split('/')[-1]
+            return (self.config['home_dir'] + '/' + file_name + '.conflict')
```

Comment:
.conflict files are not shared. We believe that a single local copy is sufficient and prevents situation where autosave would create a bunch of local copies. We can discuss this if you would like

Reply:
I just realized that you put it in the home folder of the user, right? To me this doesn't make sense. The conflict should appear in the folder of the original file, side by side. And autosave should not create a bunch of copies: if we put these checks in the save mechanism, I expect that a failure to save from the backend will cause an error message in the UI. This error message should prevent changes and stop the autosave (if it doesn't work like this, we should make it so).
Btw, since Jupyter checks if the file changed, maybe there will be some issues there... To be seen (I'm keen in proposing changes there in the upstream).

Reply:
Maybe we could store those files in something like /home/MyShares/Temp ? or just /home/Temp .

If a user started making a lot of changes then autosave would kick in multiple times and create a lot of copies, so I still think 1 .conflict file per shared file should be enough. In the same way that we usually don't create a lot of copies when we normally work on a file. 

There is another problem with saving all .conflict files in one directory though. If there are multiple files with the same name but located in different directories then it would create conflicts in the Temp directory.  We should definitely store information about original path in the file name or metadata somehow. Adding date to the file name doesn't solve this because we would still have to be able to locate the latest .conflict file for a given path.

Reply:
Actually our implementation at CERN for our different Offices, I think we create multiple files. Because we assume that (unless there's a bug), the front end should stop as soon as a save fails. I suggest we test with jupyter to see its behavior: if it shows an error to the user (blocking the possibility to continue editing) and therefore no longer does any more autosave, then we don't care about just creating a single file.

I would create the conflicts in the same dir of the file, as I said before. I will discuss it internally if we have any idea what to do with a single file share (this affects the other applications so we need to decide on a strategy either way). But I think we can assume for now that you store in the home of the user _only_ when it's a single file share. And in that case just put <shareid>_<maybe_some_path_of_a_folder_encoded?>_<filename>.conflict.<timestamp (to be unique)>.<ext>